src/storage/src/source/timestamper.rs

```diff
@@ -0,0 +1,325 @@
+use mz_expr::PartitionId;
+use mz_ore::now::NowFn;
+use mz_persist_client::read::{ListenEvent, ListenStream, ReadHandle};
+use mz_persist_client::write::WriteHandle;
+use mz_persist_client::Upper;
+use mz_repr::{Diff, Timestamp};
+
+pub struct CreateSourceTimestamper {
+    name: String,
+    read_progress: Antichain<Timestamp>,
+    write_upper: Antichain<Timestamp>,
+    persisted_timestamp_bindings: HashMap<PartitionId, VecDeque<(Timestamp, MzOffset)>>,
+    write_handle: WriteHandle<(), PartitionId, Timestamp, Diff>,
+    read_handle: ReadHandle<(), PartitionId, Timestamp, Diff>,
+    timestamp_bindings_listener: ListenStream<(), PartitionId, Timestamp, Diff>,
+    now: NowFn,
+    waker: Waker,
+}
+
+impl CreateSourceTimestamper {
+    pub async fn initialize(
+        name: String,
+        CollectionMetadata {
+            persist_location,
+            timestamp_shard_id,
+            tx_timestamp_shard_id: _,
+        }: CollectionMetadata,
+        now: NowFn,
+        waker: Waker,
+    ) -> anyhow::Result<Option<Self>> {
+        let persist_client = persist_location
+            .open()
+            .await
+            .with_context(|| "error creating persist client")?;
+
+        let mut persisted_timestamp_bindings = HashMap::new();
+
+        let (mut write_handle, read_handle) = persist_client
+            .open(timestamp_shard_id)
+            .await
+            .expect("persist handles open err");
+
+        // Collection is closed.  No source to run
+        if read_handle.since().is_empty() || write_handle.upper().is_empty() {
+            return Ok(None);
+        }
+
+        let read_progress = read_handle.since().clone();
+
+        // If this is the initialization of the persist collection, allow a read at Timestamp::minimum().  Else fetch
+        // the current upper.
+        let min_plus_one = Antichain::from_elem(Timestamp::minimum() + 1);
+        let empty: [(((), PartitionId), Timestamp, Diff); 0] = [];
+        let write_upper = write_handle
+            .compare_and_append(
+                empty,
+                Antichain::from_elem(Timestamp::minimum()),
+                min_plus_one.clone(),
+            )
+            .await
+            .expect("Initial CAA")
+            .with_context(|| "Initial compare and append")?
+            .err()
+            .map(|Upper(actual_upper)| actual_upper)
+            .unwrap_or(min_plus_one);
+
+        assert!(
+            timely::PartialOrder::less_equal(&read_progress, &write_upper),
+            "{:?} PARTIAL ORDER: {:?} {:?}",
+            name,
+            read_progress,
+            write_upper,
+        );
+
+        let mut snap_iter = read_handle
+            .snapshot(read_progress.clone())
+            .await
+            .unwrap_or_else(|e| panic!("{:?} Read snapshot at handle.since ts {:?}", name, e));
+
+        loop {
+            match snap_iter.next().await {
+                None => break,
+                Some(v) => {
```

Comment:
This can be written more concisely as `while let Some(v) = snap_iter.next().await { .. }`

Reply:
whoops yup -- "legacy code"